table/merging_iterator.cc

```diff
@@ -125,6 +120,56 @@ using MergerMinIterHeap = BinaryHeap<HeapItem*, MinHeapItemComparator>;
 using MergerMaxIterHeap = BinaryHeap<HeapItem*, MaxHeapItemComparator>;
 }  // namespace
 
+// MergingIterator uses a min/max heap to combine data from point iterators.
+// Range tombstones can be added and keys covered by range tombstones will be
+// skipped.
+//
+// The following are implementation details and can be ignored by user.
+// For merging iterator to process range tombstones, it treats the start and end
+// keys of a range tombstone as two keys and put them into minHeap_ or maxHeap_
+// together with regular point keys. Each range tombstone is active only within
+// its internal key range [start_key, end_key). An `active_` set is used to
+// track levels that have an active range tombstone. Take forward scanning
+// for example. Level j is in active_ if its current range tombstone has its
+// start_key popped from minHeap_ and its end_key in minHeap_. If the top of
+// minHeap_ is a point key from level L, we can determine if the point key is
+// covered by any range tombstone by checking if there is an l <= L in active_.
+// The case of l == L also involves checking range tombstone's sequence number.
+//
+// The following (non-exhaustive) list of invariants are maintained by
+// MergingIterator during forward scanning. After each InternalIterator API,
+// i.e., Seek*() and Next(), for all levels i, if minHeap_ is not empty:
+// 1) current_ = minHeap_.top()
+// 2) minHeap_.top().type == ITERATOR
+// 3) minHeap_.top()->key() is not covered by any range tombstone.
+//
+// After each call to SeekImpl() and FindNextVisibleKey() besides public
+// InternalIterator APIs:
+// 4) children_[i] is in minHeap_ iff children_[i].iter.Valid(), and at most
+// one children_[i] is in minHeap_.
+// 5) pinned_heap_item_[i] is in minHeap_ iff
+// range_tombstone_iters_[i]->Valid() and at most one pinned_heap_item_[i] is
+// in minHeap_.
+// When referring to range tombstone start/end keys, assume it is the value set
+// to HeapItem::parsed_ikey. This value has op_type = kMaxValid, which makes
+// range tombstone keys have distinct values from point keys.
+// 6) If range_tombstone_iters_[i]->Valid(), pinned_heap_item_[i].parsed_ikey
+// is equal to range_tombstone_iters_[i]->start_key() when type is
+// DELETE_RANGE_START and range_tombstone_iters_[i]->end_key() when type is
+// DELETE_RANGE_END (ignoring op_type which is kMaxValid for all
+// pinned_heap_item_).
+// 7) heap property is maintained: parent is always <= child in minHeap_.
+// The value used for ordering is children_[i].iter.key() for each children_[i],
+// and pinned_heap_item_[i].parsed_ikey) for each pinned_heap_item_[i].
+//
+// Define LevelNextVisible(i, k) to be the first key >= k in level i that is
+// not covered by any range tombstone.
+// 8) For all j <= i, range_tombstone_iters_[j] is at or before the first
+// range tombstone from level j that has end_key() > children_[j].iter.key().
```

Comment:
I think you meant `children_[i].iter.key()`

Reply:
Does "at or before the first range tombstone from level j that has end_key() > X" simplify to "at a range tombstone from level j that has start_key() <= X"?

Reply:
> I think you meant children_[i].iter.key()

Yes, updated.